source/lac/sparsity_tools.cc

```diff
@@ -39,15 +39,258 @@ extern "C"
+      for (int i=0; i < num_obj; ++i)
+        {
+          //Loop through each column to find neighbours
+          for ( SparsityPattern::iterator col=graph->begin(i);
+                col < graph->end(i); ++col )
+            //Ignore diagonal entries. Not needed for partitioning.
+            if ( i != col->column() )
+              {
+                *nextNborGID++ = col->column();
+                *nextNborProc++ = 0;    //All the vertices on processor 0
+              }
+        }
+    }
+
+
+    void partition_zoltan (const SparsityPattern     &sparsity_pattern,
+                           const unsigned int         n_partitions,
+                           std::vector<unsigned int> &partition_indices)
+    {
+      //MPI environment must have been initialized by this point.
+      Zoltan *zz = new Zoltan(MPI::COMM_SELF);
+
+      //Check if zoltan object is initialized
+      Assert( zz != nullptr , ExcEmptyObject() );
+
+      //General parameters
+      zz->Set_Param( "LB_METHOD", "GRAPH" );  //graph based partition method (LB-load balancing)
+      zz->Set_Param( "DEBUG_LEVEL", "0" );   //set level of debug info
+      zz->Set_Param( "NUM_LOCAL_PARTS", std::to_string(n_partitions) ); //set number of partitions
+
+      //Need a non-const object equal to sparsity_pattern
+      SparsityPattern graph;
+      graph.copy_from(sparsity_pattern);
+
+      //Set query functions
+      zz->Set_Num_Obj_Fn(get_number_of_objects, &graph);
+      zz->Set_Obj_List_Fn(get_object_list, &graph);
+      zz->Set_Num_Edges_Multi_Fn(get_num_edges_list, &graph);
+      zz->Set_Edge_List_Multi_Fn(get_edge_list, &graph);
+
+      //Variables needed by partition function
+      int changes=0;
+      int num_gid_entries=1;
+      int num_lid_entries=1;
+      int num_import=0;
+      ZOLTAN_ID_PTR import_global_ids=nullptr;
+      ZOLTAN_ID_PTR import_local_ids=nullptr;
+      int *import_procs=nullptr;
+      int *import_to_part=nullptr;
+      int num_export=0;
+      ZOLTAN_ID_PTR export_global_ids=nullptr;
+      ZOLTAN_ID_PTR export_local_ids=nullptr;
+      int *export_procs=nullptr;
+      int *export_to_part=nullptr;
+
+      //call partitioner
+      int rc = zz->LB_Partition (changes,
+                                 num_gid_entries,
+                                 num_lid_entries,
+                                 num_import,
+                                 import_global_ids,
+                                 import_local_ids,
+                                 import_procs,
+                                 import_to_part,
+                                 num_export,
+                                 export_global_ids,
+                                 export_local_ids,
+                                 export_procs,
+                                 export_to_part);
+
+      //check for error code in partitioner
+      Assert( rc == ZOLTAN_OK , ExcEmptyObject() );
+
+      Assert (partition_indices.size() != 0, ExcZero());
```

Comment:
This would indicate an error in Zoltan. I would map that to `ExcInternalError`.

Reply:
it's actually input to the function. We already have

```
Assert (partition_indices.size() == sparsity_pattern.n_rows(),
            ExcInvalidArraySize (partition_indices.size(),
                                 sparsity_pattern.n_rows()));
```

so this one can simply go.